File: src/GameImpl/Maze.fs
namespace BotRace.Game.Implementation.FSharp
open System
open BotRace.Game

    module internal _Maze =
        open _Cell
        open _Position

        type T = _Cell.T [][]

        let Create size =  
            Array.init size 
                (fun i -> 
                    Array.init size
                        (fun i ->_Cell.Create()))

        let CellAt position (maze : T) =
            maze.[position |> Row].[position |> Column]

        let Height maze = 
            Array.length maze

        let Width (maze : T) = 
            Array.length maze.[0]

//        let ToJson (maze : T) = 
//            let grid = new System.Text.StringBuilder()
//            maze.Grid 
//                |> Array2D.iter 
//                    (fun cell -> 
//                        (cell |> ToJson |> grid.Append) |> ignore)
//            grid.ToString()

        let ToJson (maze : T) = 
            maze 
                |> Array.map 
                    (fun row -> 
                        row 
                        |> Array.map (fun cell -> 
                                        cell |> ToJson )
                        |> String.Concat)
                |> String.Concat

        let Home maze : _Position.T = 0,0
        let Exit maze : _Position.T = Height maze, Width maze

        let private Update (m: T)  (p: _Position.T) c = 
            m.[p |> Row].[p |> Column] <- c

        let Carve m from direction = 
            let ``to`` = from |> At direction
            m 
            |> CellAt from
            |> Carve direction
            |> Update m from
            m
            |> CellAt ``to``
            |> Carve (direction.Oposite())
            |> Update m ``to``

        let Exists maze position = 
            let size = Array.length maze
            let col = position |> Column
            let row = position |> Row
            0 <= col && col < size && 0 <= row && row < size

 
type Maze (data : _Maze.T) =
    let maze = data
    new (size : int) = new Maze(_Maze.Create(size))

    static member ClosedGrid (size : int) =
        new Maze(size)

    member this.Carve (from : Position) direction = 
        _Maze.Carve maze (from.Row, from.Column) direction

    interface BotRace.Game.Maze with
        member this.get_Width()  = _Maze.Width maze
        member this.get_Height() = _Maze.Height maze
        member this.CellAt position = 
            maze
            |> _Maze.CellAt (position.Row, position.Column)
            |> Implementation.FSharp.Cell
            :> Cell
        member this.Home = 
            maze 
            |> _Maze.Home
            |> Implementation.FSharp.Position
            :> Position
        member this.Exit = 
            maze
            |> _Maze.Exit
            |> BotRace.Game.Implementation.FSharp.Position
            :> BotRace.Game.Position
        member this.ToJson() = maze |> _Maze.ToJson

File: src/Implementations/CSharp/GameImpl/Maze.cs
using System.Collections.Generic;
using System.Text;

using ICell = BotRace.Game.Cell;

namespace BotRace.Game.Implementation.CSharp
{
    public class Maze : BotRace.Game.Maze
    {
        private Cell[][] Grid { get; set; }

        internal static Maze ClosedGrid(int size)
        {
            var grid = new List<Cell[]>();

            for (int row = 0; row < size; row++)
            {
                var cellsRow = new List<Cell>();

                for (int column = 0; column < size; column++)
                {
                    cellsRow.Add(new Cell());
                }

                grid.Add(cellsRow.ToArray());
            }

            return new Maze { Grid = grid.ToArray() };
        }

        internal bool Exists(BotRace.Game.Position position)
        {
            var size = Grid.Length;

            return 0 <= position.Column && position.Column < size && 
                   0 <= position.Row && position.Row < size;
        }

        public int Width { get { return Grid[0].Length; } }
        
        public int Height { get { return Grid.Length; } }

        public ICell CellAt(BotRace.Game.Position position)
        {
            return Grid[position.Row][position.Column];
        }

        public string ToJson()
        {
            var grid = new StringBuilder();

            foreach (var row in Grid)
            {
                foreach (var cell in row)
                {
                    grid.Append(cell.ToJson());
                }
            }


            return grid.ToString();
        }

        public BotRace.Game.Position Home 
        {
            get
            {
                return new Position(0, 0);
            }
        }

        public BotRace.Game.Position Exit
        {
            get
            {
                return new Position(Height, Width);
            }
        }

        internal void Carve(Position from, Direction direction)
        {
            var fromCell = CellAt(from);
            var toCell = CellAt(from.At(direction));

            fromCell.Carve(direction);
            toCell.Carve(direction.Oposite());
        }
    }
}
